pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java

```diff
@@ -0,0 +1,53 @@
+package org.pac4j.kerberos.credentials.extractor;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import org.pac4j.kerberos.credentials.KerberosCredentials;
+import org.pac4j.core.context.HttpConstants;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.extractor.CredentialsExtractor;
+import org.pac4j.core.exception.CredentialsException;
+import org.pac4j.core.exception.DeferredHttpAction;
+import org.pac4j.core.exception.DeferredHttpActionCallback;
+
+/**
+ * To extract Kerberos headers.
+ *
+ * @author Garry Boyce
+ * @since 1.9.1
+ */
+public class KerberosExtractor implements CredentialsExtractor<KerberosCredentials> {
+    private final String clientName;
+
+    public KerberosExtractor(final String clientName) {
+        this.clientName = clientName;
+    }
+
+    @Override
+    public KerberosCredentials extract(WebContext context) {
+        final String header = context.getRequestHeader(HttpConstants.AUTHORIZATION_HEADER);
+        if (header == null) {
+            throw DeferredHttpAction.deferredHttpAction("Kerberos Header not found", new DeferredHttpActionCallback() {

```

Comment:
I'm not following you here.. I don't want the negotiate to occur immediately if the header is not found. In my case the caller needs to decide if the negotiate code should run and when it should run. All the current methods in HttpAction perform the action immediately rather than on demand. Maybe you can explain your thinking a bit regarding how I can get the client to throw the exception saying that the artifact it needs isn't there but that if you need it I, the client, support a method to negotiate it. My use cases are as follows: 1) If there is an authentication failure, it's not a badcredentialsexception and the client supports negotiation then it should run 2) if it is a direct  client, if there are is only one direct client in play and that client supports negotiate then perform negotiation 3) if it's a direct client used in an indirect flow then perform the negotiate where you would normally perform a redirect